web-common/src/features/dashboards/state-managers/loaders/most-recent-explore-state.ts

```diff
@@ -0,0 +1,97 @@
+import { validateAndCleanExploreState } from "@rilldata/web-common/features/dashboards/stores/validate-and-clean-explore-state";
+import { DashboardState_ActivePage } from "@rilldata/web-common/proto/gen/rill/ui/v1/dashboard_pb";
+import type {
+  V1ExploreSpec,
+  V1MetricsViewSpec,
+} from "@rilldata/web-common/runtime-client";
+import type { MetricsExplorerEntity } from "../../stores/metrics-explorer-entity";
+
+function getKeyForLocalStore(
+  exploreName: string,
+  storageNamespacePrefix: string | undefined,
+) {
+  return `rill:app:explore:${storageNamespacePrefix ?? ""}${exploreName}`.toLowerCase();
+}
+
+export function getMostRecentPartialExploreState(
+  exploreName: string,
+  storageNamespacePrefix: string | undefined,
+  metricsViewSpec: V1MetricsViewSpec,
+  exploreSpec: V1ExploreSpec,
+) {
+  const key = getKeyForLocalStore(exploreName, storageNamespacePrefix);
+  try {
+    const rawExploreState = localStorage.getItem(key);
+    if (!rawExploreState) {
+      // Return this so that destructuring is simple
+      return { mostRecentPartialExploreState: undefined, errors: [] };
+    }
+
+    const stateFromLocalStorage = JSON.parse(
+      rawExploreState,
+    ) as Partial<MetricsExplorerEntity>;
+    const errors = validateAndCleanExploreState(
+      metricsViewSpec,
+      exploreSpec,
+      stateFromLocalStorage,
+    );
+    return { mostRecentPartialExploreState: stateFromLocalStorage, errors };
+  } catch {
+    // no-op
+  }
+  // Return this so that destructuring is simple
+  return { mostRecentPartialExploreState: undefined, errors: [] };
+}
+
+export function saveMostRecentExploreState(
+  exploreName: string,
+  storageNamespacePrefix: string | undefined,
+  exploreState: MetricsExplorerEntity,
+) {
```

Comment:
Here's another case– it feels like there are two function layers (`saveMostRecentExploreState`, `setMostRecentExploreState`) where there should just be one layer. The near-identical function signatures is the giveaway.

Reply:
This follows your comment in an earlier PR. `setMostRecentExploreState` is a method used in test as well. Lets keep them separate

Reply:
Ok. +1 to unifying the various state management tests into one place.